Granda Mento smart contract implementation

tkporter opened the pull request:
### Description

For more info on the implementation, check out the descriptions of #8019 and #8082, which are PRs with smaller features that were made to the `granda-mento` branch. This PR is to merge those changes in the `granda-mento` branch into `master`.

Includes:
* `GrandaMento.sol` feature complete according to [CIP 38](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0038.md)
* `IntegerLinkedList.sol`, which is a thin wrapper of a library around `LinkedList.sol` for uint256s
* Changes `StableToken.sol` and `StableTokenEUR.sol` to allow `GrandaMento` to mint/burn stable tokens

### Other changes

* Some mild modifications to mock contracts used by tests:
   * `MockSortedOracles` was using a denominator for fixed point numbers that we moved away a long time ago when we moved to fixidity
   * Beefed up the functionality of `MockStableToken` and `MockGoldToken`. Changed some relevant tests to work with the changes.

### Tested

Unit tests

### Related issues

- Fixes #8128

### Backwards compatibility

Backward compatible

### Documentation

-

Commit:
upgrade AS and ODIS dockerfiles to node v12 (#7955)

### Description

Bumps Attestation Service and ODIS dockerfiles to node 12. This change is required due to node 10 end of life. 

### Other changes

None

### Tested

Not tested

### Related issues

- Fixes https://app.zenhub.com/workspaces/cap-sprint-board-60059934ed9c6f000e5594f6/issues/celo-org/celo-monorepo/7930

### Backwards compatibility

Yes

### Documentation

None

eelanagaraj committed:
Fix local CLI tests: dynamically import TransportNodeHid (#7959)

### Description

`yarn run test` did not run locally previously due to an issue with `jest` and importing `@ledgerhq/hw-transport-node-hid`. Dynamically importing this package (i.e. only when the `--useLedger` flag is set) fixes the issue.

**Note: this fixes the jest issue but some tests locally fail (run, but fail) while they pass on CI. Looking into this.**

Closes #6778

### Testing
- running tests locally previously failed before instantiating test cases (all but the first test run locally would fail due to the same ledger import error), and this now actually runs the test cases
- manually tested that few commands with `--useLedger` still work (`celocli account:list --useLedger --ledgerAddresses 1`, `celocli transfer:gold ... --useLedger`, `celocli exchange:dollars ... --useLedger`)

pedro-clabs committed:
Update oracle parameters. (#7920)

### Description

Replace EXCHANGES by PRICE_SOURCES and MID_AGGREGATION_{ASK,BID}_MAX_PERCENTAGE_DEVIATION by MID_AGGREGATION_MAX_PERCENTAGE_DEVIATION. The oracles are expected to run producing the same values as before, but using the new implicit pairs code.

### Tested

This configuration has been tested on Baklava.

### Related issues

- https://github.com/celo-org/celo-oracle/issues/12
- https://github.com/celo-org/celo-oracle/issues/13
- https://github.com/celo-org/celo-oracle/issues/14

### Backwards compatibility

The oracle changes are not backwards compatible, and this PR updates the deployment to use the oracle's new configuration parameters.

pedro-clabs committed:
Add a Binance CELO/EUR implicit pair.

Via USDT.

critesjosh committed:
fix broken link to mobile directory (#8020)

Commit:
Restored blockscout segment key secret (#7949)

### Description

It seems that in [this commit](https://github.com/celo-org/celo-monorepo/commit/3d2620cee47befa18de0f1329ae9713224ad77b6#diff-3de6b6ef7cbff719dc160186a180cfb760d9ca3f72946fb8c950b093326652dc) the `BLOCKSCOUT_SEGMENT_KEY` secret got lost. Most likely because the encoded file hadn't been decoded before adding new secrets.

This PR restores the secret.

### Tested

Tested on alfajores. @alecps @kevjue please verify that I haven't messed up your secrets.

Commit:
Bump WalletConnect version (#8012)

### Description

Minor code changes needed but nothing huge that breaks packages dependent on us. Big reductions in bundle size.

### Tested

With the E2E tests and unit tests.

### Backwards compatibility

N/A

### Documentation

N/A

jcortejoso committed:
Using mycelo for deploying testnets (#7803)

### Description

Introducing `mycelo` as an option for deploying a testnet. If the new env variable `GETH_USE_MYCELO` equals to `true`, mycelo will be used to generate the genesis file for the network.
Currently a limited number of configuration can be used to configure the network, like the mnemonic, number of validators, fauceted accounts (`LOAD_TEST_CLIENTS`), blocktime, epoch period and network gas limit. A further PR celo-blockchain#1390 will introduce more  customization from the migrationConfig.js. It is also possible to modify the governable parameters once the network is deployed.

### Other changes

The genesis file sharing among nodes is done using a gcp bucket. Genesis file with migrations is far larger that Kubernetes configmap limit size (around 1MB). We could consider using a local file artifact as minio if we prefer to remove gcp dependency.

### Tested

Deployed testnet

### Related issues

- Fixes celo-blockchain#1383

### Backwards compatibility

No problem. Requires access to the gcp bucket (public bucket)

### Documentation

_The set of community facing docs that have been added/modified because of this change_

eelanagaraj committed:
Fix governance:build-proposal bigNumber bug (#8028)

Looks like this was related to https://github.com/ethers-io/ethers.js/issues/228, but we just need to force the bignumber string to not be in scientific notation when transforming the value, and then this works as expected. Note that the outputted value in the outputted transaction will **not** be in scientific notation.

Closes #7466

mcortesi committed:
E2E Test Improvements (#8036)

### Description

Collection on small changes:

 * Use mycelo for e2e tests
 * Replace deprecated geth flags by the new ones
 * Fix flakyness due to concurrent startup of geth nodes

pedro-clabs committed:
Add token address support to oracle commands. (#8010)

* Add token address support to oracle commands.

* Update documentation.

Commit:
Add CELOBTC oracles config for Alfajores (#8041)

jcortejoso committed:
Fixed nginx ip address for contexted and regional clusters (#8051)

### Description

Fix the ingress-nginx external address registration for contexted clusters when region is not the default gcp region.
Also modified the nginx autoscaling parameters to be closer to current usage in mainnet cluster (current setup has always max replicas because the default memory usage for nginx controller pods is over 50% of requested.

eelanagaraj committed:
Add manual personal_decrypt flow to WalletConnect CLI client (#8037)

- Adds a `personal_decrypt` flow (based on WC unit tests)
- Adds back some import statements (commented out) for commented out flows
- Some nits (removed redundant comments)

Commit:
Granda Mento - exchange proposal implementation (#8019)

* Scaffolding

* Add extra test case

* Scaffolding

* Add extra test case

* comment change

* remove unnecessary migrations config

* add releaseData

* A lot

* Add spread to initializer

* better test case wording

* Working tests

* tests in a decent spot

* clean up

* update initialization params in release4.json

* Fix lint

* Create MockToken to fix attestations tests

* Update GrandaMento.sol

* Rename Empty state to None

* Rename proposeExchange -> createExchangeProposal

* PR comments

* fix lint

* Fix attestations test

Commit:
Granda Mento - exchange proposal implementation (#8019)

* Scaffolding

* Add extra test case

* Scaffolding

* Add extra test case

* comment change

* remove unnecessary migrations config

* add releaseData

* A lot

* Add spread to initializer

* better test case wording

* Working tests

* tests in a decent spot

* clean up

* update initialization params in release4.json

* Fix lint

* Create MockToken to fix attestations tests

* Update GrandaMento.sol

* Rename Empty state to None

* Rename proposeExchange -> createExchangeProposal

* PR comments

* fix lint

* Fix attestations test

Commit:
tobikuhlmann/update locked gold rewards docs (#8059)

### Description

This PR adds some details of the voting yield adjustment factor mechanism to the epoch rewards documentation, specifically to the Rewards for Locked Gold Holders page. The documentation has been a little sparse so far and I added some details here.

### Other changes

-

### Tested

-

### Related issues

-

### Backwards compatibility

It's just some additional lines of text to the documentation so should be fully backwards compatible. 

### Documentation

- celo-monorepo/packages/docs/celo-codebase/protocol/proof-of-stake/locked-gold-rewards.md

Commit:
Merge branch 'granda-mento' of github.com:celo-org/celo-monorepo into granda-mento

Commit:
Make updateMembershipHistory in Validators.sol correct if epoch number is 0 (#8060)

### Description

Celo-blockchain has a tool called mycelo which applies the core contract migrations as part of genesis generation. This means that the epoch number when they are run is zero. However, the updateMembershipHistory() method in Validators.sol did not expect that to be possible. As a result, when mycelo adds a validator to a validator group, the method does not add a new entry to the member's history (specifically, it does not increase numEntries from 0 to 1 as it should). This commit fixes that by adding a condition to the check for whether to replace an existing entry, namely that the number of entries isn't zero.

This will allow the e2e slashing tests to use mycelo instead of running the migrations against the live network.

### Other changes

* Updated the contract to use `InitializableV2` as required by CI.

### Tested

* Verified that, before this fix, validators' membership history is empty when they are added to the group using mycelo
* Verified that, after this fix, validators' membership history is correct when they are added to the group using mycelo
* Protocol package's automated tests pass
* Logically speaking, the change makes sense

### Backwards compatibility

Backwards compatible, as this only fixes a bug in an edge case the contract wasn't originally designed to handle.

jcortejoso committed:
Migrate Forno instances to Kubernetes.  (#8029)

### Description

This PR includes different changes needed for the migration from geth full-nodes VM based to K8s Pods. In particular:
- Archive fullnodes moved from VMs to K8s in mainnet.
- Added a new context `GCP_PRIVATE_TXNODES` for the archive full-nodes cluster in mainnet. This context is deployed to the top-level cluster with release name `rc1-fullnodes` (a forward PR could include the capability of customizing the celo-fullnode deployment name, allowing the deployment of multiple fullnode chart releases in the same cluster).
- Added the contexted variables :

1. `<CTX>_FULL_NODES_RPC_API_METHODS`: Customize the rpc and ws allowed methods in the celo-fullnodes.
2. `<CTX>_FULL_NODES_GETH_GC_MODE`: Customize geth gc mode.
3. `<CTX>_FULL_NODES_USE_GSTORAGE_DATA`: Enable the chain restoring option or not. Currently there is not available backup for archive nodes.

- Modified blockscout setup to allow custom rpc/ws endpoints using the variables `BLOCKSCOUT_OVERRIDE_RPC_ENDPOINT` and `BLOCKSCOUT_OVERRIDE_WS_ENDPOINT`. This is needed because archive nodes setup in mainnet is not based on `testnet` helm package but `celo-fullnode` helm package.
- Updated forno terraform module to terraform>=1.13.
- Added `rc1-forno.celo-testnet.org.` as domain to (new)forno domains.
-  Added `FORNO_BANNED_CIDR` to ban CIDRs to use forno.

### Other changes

- Adding a k8s service to prometheus setup.
- Updated some docker images to node:12.

### Tested

Most of the changes are already deployed.

### Related issues

- Fixes celo-blockchain#1393

### Backwards compatibility

Only compatibility problem I can think about is if someone is using old forno ip directly. 

### Pending work

- [x] Update the env files to add the new contexted variables to the existing contexts
- [x] Apply the terraform changes. 
- [ ] Update blockscout deployment to start using the k8s archive nodes

### Documentation

_The set of community facing docs that have been added/modified because of this change_

jcortejoso committed:
Added prestop hook to fullnode template. Increased baklava memory (#8095)

### Description

Added [lifecycle preStop](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/) hook to reduce the possibilities of an ungracefully shutdown of geth pods if k8s decides to kill them (because they are rescheduled, the pod is using more resources that allowed, etc...).

Also increased the baklava pod memory assigned and removed the limit (does not affect the cluster size with our node setup as cpu is the more limited resource and the ratio in current nodes is more than 1:2 for cpu cores:gb memory, so there are a lot of unassigned memory).

Upgrade already applied in baklava.

Commit:
WalletConnect tutorial to docs sidebar (#8094)

### Description

Add WC tutorial to the docs sidebar. 
### Other changes

N/A

### Tested

N/A

Commit:
Upgrade vulnerable dependencies (#8073)

* Upgrade lodash dependency

* Upgrade web3 dependency

* remove patch bn.js

* Upgrade dependency

* Upgrade sdk to 1.3.6

* Upgrade wallets to 1.3.6

* Upgrade web3-eth-contracts

* Commit yarn.lock

* Upgrade normalize-url

Commit:
WalletConnect documentation type (#8097)

Small typo that make the link not render correctly.

YazzyYaz committed:
doc: Update Cohort 7 info (#8093)

Adds Cohort 7 Information for Foundation Voting Policy

Commit:
Remove gas price sections from docs for using Metamask. (#8103)

When CIP35 was being developed, Metamask did not properly handle setting a gas price for networks other than Ethereum's mainnet.  Since then, however, they have fixed this by querying the node's eth_gasPrice method.  Since this is how we were recommended developers choose a gas price, it is not redundant as Metamask does it by itself.

asaj committed:
Integration test for adding a new StableToken (#7179)

### Description

This PR adds an integration test for adding a new StableToken, mimicking the planned process for deployment of cEUR, and future additional stable assets.

### Other changes

None.

### Tested

These are tests!

### Related issues

- Fixes #7224 

### Backwards compatibility

Yes, just tests.

Commit:
Account utilities for normalization and heuristic based correction of mnemonic phrases (#8034)

### Description

Adds improved normalization for BIP-39 mnemonic phrases, as well as a new heuristic-based method for generating suggested corrections to a malformed mnemonic phrase.

Autocorrect suggestions are designed to correct words that may have been copied incorrectly, or mistyped when entered to recover an account. It suggests similar phrases, ordered by estimated probability that they are the source phrase of the observed phrase after being passed through a "noisy channel". Ordering is determined by edit total edit distance of the observed (incorrect) phrase to the suggestion.

### Other changes

* Major refactoring of `accounts.test.ts`
* Refactoring of `accounts.ts`

### Tested

Added unit tests which include a number of new phrases for normalization and correction testing.

### Related issues

- Related https://github.com/celo-org/celo-monorepo/issues/7060

codyborn committed:
Add Valora-Accounts to nav page (#8119)

### Description

Minor change to add this existing doc to the nav bar

codyborn committed:
Use non-US for Nexmo default (#8067)

### Description

Per 10DLC requirement, we'll begin asking validators to register a US toll-free number in addition to a non-US number. This US toll-free number can only be used within US and Canada and therefore we should default the sending phone number to a non-US number if we don't have a country code match. This is only important for Nexmo since Twilio has intelligent routing.

### Tested

Tested using a live validator

### Related issues

- Fixes #7937 

### Backwards compatibility

Yes

codyborn committed:
Add more params to status endpoint (#8125)

Commit:
Granda Mento: support for approving, cancelling, and executing exchange proposals (#8082)

* Scaffolding

* Add extra test case

* Scaffolding

* Add extra test case

* comment change

* remove unnecessary migrations config

* add releaseData

* A lot

* Add spread to initializer

* better test case wording

* Working tests

* tests in a decent spot

* clean up

* update initialization params in release4.json

* Fix lint

* Make contract changes, now doing tests

* Create MockToken to fix attestations tests

* Update GrandaMento.sol

* Rename Empty state to None

* Tests working

* Add setApprover test

* Tests and cancel working

* Rename proposeExchange -> createExchangeProposal

* Rename proposeExchange -> createExchangeProposal

* Allow GrandaMento to mint and burn

* wow everything works

* PR comments

* fix lint

* Use registry IDs for stable tokens

* Bump minor version of StableToken and StableTokenEUR

* Use the string registry ID rather than the keccak for exchange limits

* Create assertRevertWithReason

* Remove console.log

* Fix unit and protocol version tests

* Add constructor to StableTokenEUR

* Using IntegerLinkedList for activeProposalIds, need to change tests to have proposal IDs start with 1 instead of 0

* fix tests

* Fix stabletoken test

* Bump stable token patch version rather than minor

* better comments

* Fix test and start debugging protocol test release failure

* Fix protocol-test-release (i hope)

* Apply suggestions from code review

Co-authored-by: Martín Volpe <volpe@clabs.co>

* Modest changes from review

* Comment in tests

* Move to liquidity folder

Co-authored-by: Martín Volpe <volpe@clabs.co>

Commit:
Granda Mento: support for approving, cancelling, and executing exchange proposals (#8082)

* Scaffolding

* Add extra test case

* Scaffolding

* Add extra test case

* comment change

* remove unnecessary migrations config

* add releaseData

* A lot

* Add spread to initializer

* better test case wording

* Working tests

* tests in a decent spot

* clean up

* update initialization params in release4.json

* Fix lint

* Make contract changes, now doing tests

* Create MockToken to fix attestations tests

* Update GrandaMento.sol

* Rename Empty state to None

* Tests working

* Add setApprover test

* Tests and cancel working

* Rename proposeExchange -> createExchangeProposal

* Rename proposeExchange -> createExchangeProposal

* Allow GrandaMento to mint and burn

* wow everything works

* PR comments

* fix lint

* Use registry IDs for stable tokens

* Bump minor version of StableToken and StableTokenEUR

* Use the string registry ID rather than the keccak for exchange limits

* Create assertRevertWithReason

* Remove console.log

* Fix unit and protocol version tests

* Add constructor to StableTokenEUR

* Using IntegerLinkedList for activeProposalIds, need to change tests to have proposal IDs start with 1 instead of 0

* fix tests

* Fix stabletoken test

* Bump stable token patch version rather than minor

* better comments

* Fix test and start debugging protocol test release failure

* Fix protocol-test-release (i hope)

* Apply suggestions from code review

Co-authored-by: Martín Volpe <volpe@clabs.co>

* Modest changes from review

* Comment in tests

* Move to liquidity folder

Co-authored-by: Martín Volpe <volpe@clabs.co>

Commit:
Remove some console.logs

Commit:
Remove some console.logs

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into granda-mento

martinvol committed:
Fix integration tests (#8138)

martinvol committed:
Fix integration tests (#8138)

martinvol committed:
Fix docs for CP-DOTO (#8141)

### Description

The article about adding multiple stable tokens was not being indexed.

### Other changes

_Describe any minor or "drive-by" changes here._

### Tested

_An explanation of how the changes were tested or an explanation as to why they don't need to be._

### Related issues

- Fixes #[issue number here]

### Backwards compatibility

_Brief explanation of why these changes are/are not backwards compatible._

### Documentation

_The set of community facing docs that have been added/modified because of this change_

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+
+  // Emitted when the exchange limits for a stable token are set.
+  event StableTokenExchangeLimitsSet(
+    string stableTokenRegistryId,
```
does it make sense for the registry ID to be used here but the address to be used in `ExchangeProposalCreated`?

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+    // The exchanger/proposer of the exchange proposal.
+    address payable exchanger;
+    // The stable token involved in this proposal.
+    address stableToken;
```
ditto
ID versus address?

tkporter replied:
This one's actually important to use the address rather than the ID just in case a stable token -> CELO proposal is created, the exchanger deposits that stable token, and then the address in the registry of the stable token changes. Then when the exchange is cancelled or executed, the contract will try to burn / transfer the deposited tokens from the new address rather than the old one
Though for CELO -> stable token exchanges this gets weird in the above situation. Maybe we could store both `address stableToken` and `address stableTokenRegistryId` and either use `stableToken` for stable token -> CELO exchanges and `stableTokenRegistryId` for CELO -> stable token exchanges, or maybe just refund the deposit if the address for `stableTokenRegistryId` differs from `stableToken` at execution time?

yorhodes replied:
hmm
I think we would never want to make a storage migration where the new balances don't reflect the old balances

tkporter replied:
Given that, how do you feel about the assumption the token address will never change?

yorhodes replied:
I have imagined a pattern where we could have a new address but the new contract falls back to the old contract's state if uninitialized/migrated
I don't feel strongly about this case though

tkporter replied:
Chatted 1:1, keeping with this approach

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+    // The state of the exchange proposal.
+    ExchangeProposalState state;
+    // Whether CELO is being sold and stableToken is being bought.
+    bool sellCelo;
```
the enum and booleans should be ordered in between the `address` and `uint256` to save on storage slots
```suggestion
    // The exchanger/proposer of the exchange proposal.
    address payable exchanger;
    // The stable token involved in this proposal.
    address stableToken;
    // The state of the exchange proposal.
    ExchangeProposalState state;
    // Whether CELO is being sold and stableToken is being bought.
    bool sellCelo;
    // The amount of the sell token being sold. If a stable token is being sold,
    // the amount of stable token in "units" is stored rather than the "value."
    // This is because stable tokens may experience demurrage/inflation, where
    // the amount of stable token "units" doesn't change with time, but the "value"
    // does. This is important to ensure the correct inflation-adjusted amount
    // of the stable token is transferred out of this contract when a deposit is
    // refunded or an exchange selling the stable token is executed.
    // See StableToken.sol for more details on what "units" vs "values" are.
    uint256 sellAmount;
    // The amount of the buy token being bought. For stable tokens, this is
    // kept track of as the value, not units.
    uint256 buyAmount;
    // The timestamp (`block.timestamp`) at which the exchange proposal was approved
    // in seconds. If the exchange proposal has not ever been approved, is 0.
    uint256 approvalTimestamp;
```

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+    // Require the configurable stableToken max exchange amount to be > 0.
+    // This covers the case where a stableToken has never been explicitly permitted.
+    ExchangeLimits memory exchangeLimits = stableTokenExchangeLimits[stableTokenRegistryId];
+    require(exchangeLimits.maxExchangeAmount > 0, "Max stable token exchange amount must be > 0");
```
maybe wrap this in a public `getExchangeLimits`
and maybe instead of "must be > 0" it should be "must be explicitly permitted/defined"

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+    }
+    // The sell amount taking the spread into account, ie:
+    // (1 - spread) * sellAmount
+    FixidityLib.Fraction memory adjustedSellAmount = FixidityLib.fixed1().subtract(spread).multiply(
+      FixidityLib.newFixed(sellAmount)
+    );
```
if `spread` is in storage than maybe its worth storing `fixed1().subtract(spread)` instead
My intuition tells me there are some major inefficiencies in this function with the (quite heavy) fixidity computations

yorhodes replied:
I understand that it might not matter given the intended use of `createExchangeProposal`

tkporter replied:
My only qualm is I'm not too sure what to name a stored `fixed1().subtract(spread)` without it being confusing. Especially given `Exchange.sol` stores `spread` so there's some familiarity with that already

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+    exchangeProposalCount = exchangeProposalCount.add(1);
+    exchangeProposals[exchangeProposalCount] = ExchangeProposal({
+      exchanger: msg.sender,
+      stableToken: stableToken, // for stable tokens, is saved in units to deal with demurrage.
```
this comment looks like it's about `stableToken`

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+    exchangeProposals[exchangeProposalCount] = ExchangeProposal({
+      exchanger: msg.sender,
+      stableToken: stableToken, // for stable tokens, is saved in units to deal with demurrage.
+      sellAmount: sellCelo ? sellAmount : IStableToken(stableToken).valueToUnits(sellAmount),
```
this would be more readable if it was separate from the write to storage struct

yorhodes replied:
I still don't understand why the exchanger's proposal shouldn't be affected by `value:units` fluctuations during the lifetime of the proposal
it seems like this should be computed at execution time? I think I'm missing something

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+      state: ExchangeProposalState.Proposed,
+      sellCelo: sellCelo
+    });
+    // Push it into the array of active proposals.
```
nit: array

m-chrzan replied:
Isn't this a LinkedList, not an array?

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+    // Require the appropriate state and sender.
+    // This will also revert if a proposalId is given that does not correspond
+    // to a previously created exchange proposal.
+    require(
+      (proposal.state == ExchangeProposalState.Proposed && proposal.exchanger == msg.sender) ||
+        (proposal.state == ExchangeProposalState.Approved && isOwner()),
+      "Sender cannot cancel the exchange proposal"
+    );
```
should the approver also be able to cancel in `State.Approved`?

tkporter replied:
Hmm that's a good question. Part of the reasoning why only Governance can cancel in the `Approved` state was to prevent the exchanger from having a "free option" where they can just cancel the exchange when the price they locked in at proposal-time is worse than the price at exchange-execution-time. Giving the approver the ability to cancel when it's Approved could make it easier for the exchanger to do this if they're able to convince the approver to cancel it on their behalf. But allowing this could prevent needing to rally the community behind a governance proposal if the approvers mistakenly approved a proposal. I'm generally supportive of it, what're your thoughts?

tkporter replied:
Following the brief discussion [here](https://github.com/celo-org/celo-proposals/issues/229#issuecomment-866414656), deciding to keep as is

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+      refundToken.transfer(proposal.exchanger, refundAmount),
+      "Transfer out of refund token failed"
+    );
+    emit ExchangeProposalCancelled(proposalId, msg.sender);
```
is the `msg.sender` useful?

tkporter replied:
I was thinking it'd let someone easily see whether it was cancelled by the proposer or vetoed by governance. Though this seems easily implied so I'll remove

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+      sellToken = IERC20(stableToken);
+      // When selling stableToken, the sell amount is stored in units.
+      // Units must be converted to value when refunding.
+      sellAmount = IStableToken(stableToken).unitsToValue(proposal.sellAmount);
```
hmm I think I understand now
it's stored as units to *ensure* that the exchanger *is* affected by inflation

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+  function setSpread(uint256 newSpread) public onlyOwner {
+    spread = FixidityLib.wrap(newSpread);
+    emit SpreadSet(newSpread);
+  }
```
in the past we have required `<= fixed1` on fractions like this
it should be explicitly documented that the input is wrapped within the function (in @ dev ?)
or just shouldn't be wrapped; imo whatever makes the config values easier to read in plaintext should be prioritized

yorhodes replied:
ignore the latter half, forgot what `wrap` meant lol

yorhodes commented on packages/protocol/contracts/stability/StableToken.sol:
```diff
@@ -41,6 +41,8 @@ contract StableToken is
 
   event TransferComment(string comment);
 
+  bytes32 constant GRANDA_MENTO_REGISTRY_ID = keccak256(abi.encodePacked("GrandaMento"));
```
though this is already a pervasive problem, this is another departure from modularity in the contracts
a whitelist of identifiers in storage is the alternative

tkporter replied:
Yeah kinda torn on this one. I agree a whitelist of identifiers is much nicer but I'm not sure if it's appropriate for Granda Mento to also navigate that change given it'll expand the surface area a lot & make the contract release more complicated. Wdyt?

martinvol replied:
The release doesn't have to be complicated because the smart contract could store a default, we used that pattern for cEUR

yorhodes commented on packages/protocol/contracts/stability/test/MockStableToken.sol:
```diff
@@ -3,64 +3,78 @@ pragma solidity ^0.5.13;
-  function setNeedsRebase() external {
-    _needsRebase = true;
+  constructor() public {
+    setInflationFactor(FixidityLib.fixed1().unwrap());
```
not specific to this pr, but an "inflation factor of 1" suggests 100% inflation, not 0

yorhodes commented on packages/protocol/lib/test-utils.ts:
```diff
@@ -127,7 +127,28 @@ export const assertThrowsAsync = async (promise: any, errorMessage: string = '')
+    await promise
+    assert.fail('Expected transaction to revert')
+  } catch (error) {
+    // When it's a view call, error.message has a shape like:
+    // `Returned error: VM Exception while processing transaction: revert ${revertMessage}`
+    // When it's a transaction (eg a non-view send call), error.message has a shape like:
+    // `Returned error: VM Exception while processing transaction: revert ${revertMessage} -- Reason given: ${revertMessage}.`
+    // Therefore we try to parse the first instance of `${revertMessage}`.
```
is this true for `celo-blockchain` and `ganache`? this seems bizarre

tkporter replied:
Only ever tested with ganache, adding a comment indicating such

yorhodes commented on packages/protocol/migrations/11_grandamento.ts:
```diff
@@ -0,0 +1,34 @@
+  initializeArgs,
+  async (grandaMento: GrandaMentoInstance) => {
+    // Add as a spender of the Reserve
+    const reserve: ReserveInstance = await getDeployedProxiedContract<ReserveInstance>(
```
why make this 11 instead of n+1
the git blame might get funky as a result

yorhodes commented on packages/protocol/releaseData/initializationData/release4.json:
```diff
@@ -1 +1,3 @@
-{}
\ No newline at end of file
+{
+  "GrandaMento": ["0x000000000000000000000000000000000000ce10", "0x0000000000000000000000000000000000000000", "0", "1209600"]
```
this should actually go in `release5.json`
I know it's super confusing lol

martinvol replied:
we should also have an empty `release4.json` for future reference, right?

yorhodes commented on packages/protocol/test/liquidity/grandamento.ts:
```diff
@@ -0,0 +1,1207 @@
+        assert.equal(exchangeProposal.sellCelo, false)
+      })
+
+      it('stores the exchange proposal with the sell amount in units when the stable token inflation factor is not 1', async () => {
```
dedupe with previous test?

yorhodes commented on packages/protocol/test/liquidity/grandamento.ts:
```diff
@@ -0,0 +1,1207 @@
+        })
+      })
+
+      it('emits the ExchangeProposalCreated event with the sell amount as the stable token value when its inflation factor not 1', async () => {
```
dedupe with previous test?

critesjosh committed:
Update celo-for-eth-devs.md (#8006)

### Description

Fix formatting issue with gitbook.

yorhodes commented on packages/protocol/test/liquidity/grandamento.ts:
```diff
@@ -0,0 +1,1207 @@
+    })
+
+    describe('when called by the owner', () => {
+      beforeEach(async () => {
```
nit: dedupe and move beforeEach up 1 describe

tkporter replied:
Some of the specific cases need to do some things before creating an exchange proposal so keeping as is for now

yorhodes commented on packages/protocol/test/liquidity/grandamento.ts:
```diff
@@ -0,0 +1,1207 @@
+          )
+
+          const aliceBalanceBefore = await stableToken.balanceOf(alice)
+          await grandaMento.cancelExchangeProposal(1, { from: alice })
```
would love to see a lot of this and tests below deduped
1200 lines of tests seems like a lot for a fairly simple contract

tkporter replied:
Deduped a lot of tests-- still a pretty big file, but marginally better now

yorhodes reviewed (commented):
few questions and comments

Commit:
Changes for blockscout v.3.5.1 (#8143)

### Description

Changes needed to run the upgraded version of Blockscout (v3.5.1).

### Other changes

Adjusted request values.

### Tested

Tested on Alfajores and Mainnet.

critesjosh committed:
update local signing to align w/ donut upgrade (#8140)

### Description

The local signing problem doesn't exist anymore, since the donut hardfork. Updating this page of the docs to reflect that.

Commit:
Various PR comments

Commit:
Various PR comments

Commit:
Don't reshuffle migrations

Commit:
Don't reshuffle migrations

Commit:
Move granda mento to before governance in migrations, modest refactor to tests

Commit:
Move granda mento to before governance in migrations, modest refactor to tests

codyborn committed:
Speed up BLS combination (#8104)

### Description

BLS Signature verification in the Combiner takes 500 ms for each signature. This puts a lower limit of 3 seconds of latency per request. A large majority of the time, all signatures will be valid and we can get major efficiency gains by optimistically combining the sigs. Only upon failure would we go through each signature to identify the invalid ones and log the culprit. 

### Tested

Unit tests included. Performance tested in Alfajores.

### Related issues

- Fixes #7682 

### Backwards compatibility

Only an update to the Combiner in how it validates signatures. All APIs are the same.

### Documentation

N/A

jcortejoso committed:
Small kubernetes geth config improvements (#8135)

### Description
Small fixes tested to help with some corner cases or stressful situations:
- Removed celostats config when node has not imported the etherebase account in its keystore (this caused some  traces in the log)
- Increased Alfajores/Baklava cpu and memory requests for geth processes. May be interesting decouple this value for the different roles (validators, proxies, fullnodes, archive, ...). Archive may need more resources than others, for example.
- Increased `terminationGracePeriodSeconds` from 2 minutes to 5 minutes following the recommendations from the documentations. Geth can take a long time to commit the chain to disk when shut down, and killing ungracefully has a high risk of loosing chain data or corrupting the db.


### Tested

Changes already tested in alfajores.

m-chrzan commented on packages/protocol/contracts/common/linkedlists/IntegerLinkedList.sol:
```diff
@@ -0,0 +1,93 @@
+ * @title Maintains a doubly linked list keyed by uint256.
+ * @dev Following the `next` pointers will lead you to the head, rather than the tail.
+ */
+library IntegerLinkedList {
```
I'm wondering about the usefulness of this contract. I get not littering the main logic with `bytes32` casts, but this ends up functioning basically exactly like `LinkedList`, storing the same sort of data, while increasing gas costs for basic operations (or is the Solidity compiler smart enough to optimize all this away?). 
Solidity is a low level language, so personally I wouldn't be too afraid of a few `bytes32` casts in GrandaMento (or just having the IDs be `bytes32`?).

tkporter replied:
I don't have an especially strong take either way, I mostly created this to follow the pattern we've done for other type-specific LinkedList / SortedLinkedList libraries.

No idea if the compiler is smart enough to really help though. Fwiw because the library is all internal functions, iiuc the library will be deployed alongside GrandaMento.sol as bytecode that's JUMP'd to and not as a linked library that's delegatecall'd. So while there's probably an extra JUMP in there the gas costs probably aren't exorbitant. Sorta leaning toward keeping it the way it is but am not opposed to changing if there are strong opinions

m-chrzan commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+  // Approved state. Used for easily viewing all the active exchange proposals.
+  LinkedList.List private activeProposalIds;
+
+  // Number of exchange proposals that exist. Used for assigning an exchange
```
Wording nit: would "proposals that have ever been proposed" or similar work better? "That exist" might have the connotation of not having been cancelled/executed.

m-chrzan commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+  /**
+   * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param _registry The address of the registry.
+   * @param _spread The spread charged on exchanges.
```
Missing @param comments.

m-chrzan commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+      exchanger: msg.sender,
+      stableToken: stableToken, // for stable tokens, is saved in units to deal with demurrage.
+      sellAmount: sellCelo ? sellAmount : IStableToken(stableToken).valueToUnits(sellAmount),
+      buyAmount: buyAmount,
```
Shouldn't this be converted to units if `sellCelo` is true?

tkporter replied:
My rationale for a CELO -> stable token exchange is that calculating the stable token buy value at proposal-time means that the same value will be minted at execution-time, meaning the exchanger doesn't get hit by any demurrage on the stable token during that time period [proposal time, execution time].

For stable token -> CELO exchanges, an exchanger similarly isn't hit by demurrage during that time period [proposal time, execution time], because the value of the stable token at proposal time is used to calculate the buyAmount of CELO.

So the design is built to the favor of the exchanger, but imo it feels fair-- in the CELO -> stable token case, they aren't given the stable tokens until execution-time, so it doesn't seem fair to penalize them with demurrage until they're given the stable token. And in the stable token -> CELO case, they're choosing to "spend" that stable token at proposal-time, so it doesn't seem fair to penalize them after they decide to spend it.

Curious to hear your thoughts though. If you're on board I can add a comment about it too

m-chrzan commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+   * Proposed or Approved state.
+   * @return An array of active exchange proposals IDs.
+   */
+  function getActiveProposalIds() external view returns (uint256[] memory) {
```
Are we using a linked list only to make this function possible? I guess it's not a *huge* complexity cost, but wondering if we could do without it and just use an array.

tkporter replied:
Yeah. I also first thought of an array and to just have functions that would result in the removal of an element (like cancelling or executing an exchange) pass in the index of the element to the array, and to have something like `require(activeProposalIds[index] == proposalId)`. But in the case of a veto from governance, which would involve creating a governance proposal to call a function like `function cancelExchangeProposal(uint256 proposalId, uint256 activeProposalIdArrayIndex)`, it's possible that the active proposal id array would change so that by the time the governance proposal is able to be executed, the index in the proposal is incorrect.

The linkedlist approach is the only one I thought of that'd be O(1) for inserts/removals & not vulnerable to something like that, and while this view function is O(n), I see it being used for off-chain reads and not in a transaction via a call from another smart contract. Open to other ideas though

m-chrzan replied:
Do we need a separate data structure for just active proposals? I guess we're hitting the question of whether we should provide convenient view functions or let people manage their knowledge via something like event indexing. Because removing this would simplify the data structures in this contract and remove a dependency on a new library whose usefulness and overhead we're not sure of, I'm personally slightly leaning towards removing the list, assuming this functionality isn't something very necessary.

tkporter replied:
My only hesitation is I don't think there's a great setup on Celo right now for indexing events and exposing resulting info in a super easy way. It's pretty important for people to be able to see the active proposals and realistically at launch the only sorta user friendly way will be the CLI, and the easiest way is what we've done with governance which is just exposing a view function. The only thing that comes to mind is The Graph but I'm not sure if we're at a point where we feel confident in relying upon that for important indexing. I don't know much about The Graph but I'm also not sure how efficient it'll be to read through all create / cancel / exchanged events in history to figure out the current set of active proposals

yorhodes replied:
the indices would only change if we reuse them, right?
maybe a compromise is just to always increment the identifier given that we don't anticipate the number of granda mento proposals to grow rapidly?
I agree the complexity that comes with the linked list is really overkill for this application

tkporter replied:
>maybe a compromise is just to always increment the identifier given that we don't anticipate the number of granda mento proposals to grow rapidly?

Not sure I understand what you meant here?

m-chrzan commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+
+  /**
+   * @notice Sets the approver.
+   * @dev Sender must be owner. New approver is allowed to be address(0).
```
Is this equivalent to freezing GrandaMento until a new one is set?

tkporter replied:
It'd still allow proposals in the `Approved` state to be executed once the veto period is up.

Though I could see an argument for making the contract more explicitly "frozen" when the approver == address(0). If there's something badly wrong with the multisig, likely the first & easiest reaction would be a governance proposal or hotfix to set the approver to address(0), and to then take more time to figure out the details of a new approver multisig. We could possibly add a check in `executeExchangeProposal`: `require(approver != address(0))`, and then in `cancelExchangeProposal` we allow anyone to cancel an active exchange proposal (ie refund the deposit) if `approver == address(0)`. This would probably only be useful if the multisig was compromised and was approving exchanges it shouldn't be.

Another option is to have that freezing functionality be determined more explicitly by `Freezable` or by a `frozen` boolean. Any thoughts?

m-chrzan commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,477 @@
+  /**
+   * @notice Sets the approver.
+   * @dev Sender must be owner. New approver is allowed to be address(0).
+   * @param newApprover The new value for the spread.
```
Typo: spread -> approver.

m-chrzan reviewed (commented):
Only reviewed up to the GrandaMento contract itself so far.

Didn't see one in the CIP, and technically not related to this PR, but do we have proposed initial constitution values for the functions of this contract, especially for the Governance veto function?

Also, doesn't look like unapproved exchange proposals ever expire, remaining proposed basically forever (since even Governance can't veto until after approval). Maybe approver should also be allowed to cancel during the Proposal stage?

martinvol committed:
Merge branch 'master' into granda-mento

eelanagaraj committed:
JS keystore support (#8096)

### Description
- Implements most basic keystore functionality (encrypt, decrypt, change password) base class (+ file system for data persistence), based on `ethereumjs-wallet`
- `Keystore` abstraction should make it fairly straightforward to add additional IO formats (browser storage, DB, however valora wants to do this...)
- `KeystoreWalletWrapper` is a bare bones wrapper of a file keystore + the `LocalWallet` -- I expect this to change (+ maybe move into its own directory), but would wait until implementing attestation service/CLI-relevant changes before adding more here, in order to see what's necessary/makes sense

### Tested

- Unit tests
- Tested instantiating, signing, sending txs via the `KeystoreWalletWrapper` + keystore files
- Tested that keystore files created by geth can be copied into this keystore directory and decrypted with the same passphrase; same with the reverse (i.e. keystore files created + encrypted with the `FileKeystore` can be copied to the geth keystore directory, decrypted, and used to send txs)

### Related issues

- #7063

- related autogenerated docs (split PR so as to not clutter this one) [old file structure PR](https://github.com/celo-org/celo-monorepo/pull/8160), [docs PR new file structure](https://github.com/celo-org/celo-monorepo/pull/8185)

### Backwards compatibility

New module, no changes to existing wallets or tools.

### Remaining TODOs
- [later PR] modify KeystoreWalletWrapper as makes sense with future use cases

codyborn committed:
ODIS: Reduce error noise by not calling combine when insufficient sigs (#8194)

### Description

Small change to only call `combinePartialBlindedSignatures` when we have sufficient sigs. Without this check, `combinePartialBlindedSignatures` is called, even when user has no quota. Since `combinePartialBlindedSignatures` logs an error when there are not sufficient sigs, our logs are becoming noisy with unactionable errors.

### Other changes

Combiner version bump

### Tested

Unit tests


### Backwards compatibility

Yes

### Documentation

N/A

Commit:
Dedupe tests

Commit:
Dedupe tests

Commit:
PR comments

Commit:
PR comments

Commit:
Merge branch 'granda-mento' of github.com:celo-org/celo-monorepo into granda-mento

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into granda-mento

Commit:
Add some constitution values, lint fix

Commit:
Add some constitution values, lint fix

Commit:
cancelExchangeProposal as 0.6 constitution

Commit:
cancelExchangeProposal as 0.6 constitution

tkporter commented:
> Also, doesn't look like unapproved exchange proposals ever expire, remaining proposed basically forever (since even Governance can't veto until after approval). Maybe approver should also be allowed to cancel during the Proposal stage?

@m-chrzan The proposer is able to cancel during the Proposal stage, do you think that's not sufficient?

The only thing I could think of is if the proposer tries to strategically create an exchange proposal when the price of the buy asset is really cheap, and then they try to get it approved a while later when the price of the buy asset is higher, but their cheap price is locked in. In that case I'd hope the community & approver multisig would spot that and refuse to approve it.

Re: constitution, good point, I haven't really thought too strongly on it yet... I feel like the `cancelExchangeProposal` value shouldn't be crazy high. I just updated the constitution values in this PR, I'd like to hear your thoughts

m-chrzan commented:
> @m-chrzan The proposer is able to cancel during the Proposal stage, do you think that's not sufficient?
> 
> The only thing I could think of is if the proposer tries to strategically create an exchange proposal when the price of the buy asset is really cheap, and then they try to get it approved a while later when the price of the buy asset is higher, but their cheap price is locked in. In that case I'd hope the community & approver multisig would spot that and refuse to approve it.

Yeah, what you described is basically what I'd be worried about. I'm not sure how really feasible it is, but it's basically a long range attack where you start with a large trade that you never attempt to approve, wait until its value grows a lot, at that point you could potentially even bribe the current approvers and a large enough portion of Governance voters to let your proposal through.

Since the approver multisig is already a gatekeeper, there shouldn't be any harm in giving them explicit disapproval power. Exchanger can always repeat the same proposal in the future.

martinvol committed:
Fix migrations for granda mento (#8196)

martinvol committed:
Fix migrations for granda mento (#8196)

aaronmgdr committed:
Replace Wallet with Contract Kit in readme (#8190)

### Description

top of readme still said that monorepo contained the wallet. but it does not.

### Other changes

replace with contract kit

### Tested

Commit:
Add check for phrase length in mnemonic phrase correction (#8146)

### Description

Addressing comment in https://github.com/celo-org/wallet/pull/508#discussion_r654137527, add a check
that the number of words in the phrase is what is expected. Currently does not attmept to correct
phrases with incorrect length, but a heuristic could be added to do so.

_The set of community facing docs that have been added/modified because of this change_

Commit:
Merge branch 'granda-mento' of github.com:celo-org/celo-monorepo into granda-mento

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into granda-mento

yorhodes commented:
> Yeah, what you described is basically what I'd be worried about. I'm not sure how really feasible it is, but it's basically a long range attack where you start with a large trade that you never attempt to approve, wait until its value grows a lot, at that point you could potentially even bribe the current approvers and a large enough portion of Governance voters to let your proposal through.

we could put a governable upper limit on the variance between the approval and proposal price?

tkporter commented:
>we could put a governable upper limit on the variance between the approval and proposal price?

I like that. Giving the approver multisig the ability to cancel a `Proposed` exchange proposal doesn't necessarily prevent them from still receiving a bribe, but this more directly would

Commit:
adds AS error code documentation to README (#8203)

### Description

Adds a list of error codes and messages to the Attestation Service README to resolve ambiguity around the meaning of error responses for wallets and developers that integrate with the AS. 

### Other changes

None

### Tested

Not necessary. 

### Related issues

- https://app.zenhub.com/workspaces/cap-sprint-board-60059934ed9c6f000e5594f6/issues/celo-org/celo-monorepo/8022

### Backwards compatibility

Yes

### Documentation

Yes

Commit:
Add maxApprovalExchangeRateChange

Commit:
Add maxApprovalExchangeRateChange

Commit:
Fix lint

Commit:
Fix lint

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into granda-mento

nambrot commented on packages/protocol/contracts/stability/StableToken.sol:
```diff
@@ -270,12 +279,12 @@ contract StableToken is
+  function burn(uint256 value) external updateInflationFactor returns (bool) {
+    require(
+      msg.sender == registry.getAddressForOrDie(getExchangeRegistryId()) ||
+        msg.sender == registry.getAddressFor(GRANDA_MENTO_REGISTRY_ID),
```
I'm wondering if StableToken should have a governable limit for both Validators and Granda Mento as to how much they can mint? Maybe it's moot but it might limit the impact of a bug in both contracts?

tkporter replied:
I like the idea but am not sure if it fits within the scope of this PR, wdyt?

cangshu888 committed:
Add Tradegen project to dApp gallery (#8169)

### Description

Add description of Tradegen project and Github link to dApp gallery.

Commit:
Move to activeProposalIdsSuperset

Commit:
Move to activeProposalIdsSuperset

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into granda-mento

Commit:
Comment

Commit:
Comment

Commit:
fix lint

Commit:
Now fix lint

Commit:
Now fix lint

yorhodes commented on packages/protocol/contracts/liquidity/GrandaMento.sol:
```diff
@@ -0,0 +1,608 @@
+    FixidityLib.Fraction memory proposalRate = FixidityLib.wrap(
+      proposal.celoStableTokenExchangeRate
+    );
+    (FixidityLib.Fraction memory lesserRate, FixidityLib.Fraction memory greaterRate) = currentRate
+      .lt(proposalRate)
+      ? (currentRate, proposalRate)
+      : (proposalRate, currentRate);
+    FixidityLib.Fraction memory rateChange = greaterRate.subtract(lesserRate).divide(proposalRate);
+    require(
+      rateChange.lte(maxApprovalExchangeRateChange),
+      "CELO exchange rate is too different from the proposed price"
+    );
```
I guess we don't have to guard against this in both directions? ie if the proposer and approver move ahead with an exchange where the proposal rate is much higher than the current rate, thats probably fine?

tkporter replied:
Might be more complex to implement bc then it'd have to consider if it's a stableToken -> CELO or CELO -> stableToken exchange to divide the exchange rate by 1 or not. Thinking I'll keep it the way it is for now? Lmk if you have an issue with it

yorhodes commented on packages/protocol/test/liquidity/grandamento.ts:
```diff
@@ -0,0 +1,1104 @@
+        assertEqualBN(proposal.approvalTimestamp, latestBlock.timestamp)
+      })
+
+      for (const rateChange of [-maxApprovalExchangeRateChange, maxApprovalExchangeRateChange]) {
+        it(`tolerates ${
+          rateChange > 0 ? 'an increase' : 'a decrease'
+        } in the exchange rate since exchange proposal creation within the maximum`, async () => {
+          // The absolute max change
+          const newCeloStableTokenRate = defaultCeloStableTokenRate.times(1 + rateChange)
+          await sortedOracles.setMedianRate(stableToken.address, newCeloStableTokenRate)
+          await grandaMento.approveExchangeProposal(proposalId, { from: approver })
+        })
+      }
+
+      for (const rateChange of [
+        -maxApprovalExchangeRateChange - 0.01,
+        maxApprovalExchangeRateChange + 0.01,
+      ]) {
+        it(`reverts when ${
+          rateChange > 0 ? 'an increase' : 'a decrease'
+        } in the exchange rate since exchange proposal creation not within the maximum`, async () => {
+          // Just in excess of the max change
+          const newCeloStableTokenRate = defaultCeloStableTokenRate.times(1 + rateChange)
+          await sortedOracles.setMedianRate(stableToken.address, newCeloStableTokenRate)
+          await assertRevertWithReason(
+            grandaMento.approveExchangeProposal(proposalId, { from: approver }),
+            'CELO exchange rate is too different from the proposed price'
+          )
+        })
+      }
```
love this

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into granda-mento

Commit:
Rename var

Commit:
Merge branch 'granda-mento' of github.com:celo-org/celo-monorepo into granda-mento

Commit:
rm InitializableV2

Commit:
New registry constructor param

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into granda-mento

Commit:
Merge branch 'master' into granda-mento

mergify[bot] committed:
Merge branch 'master' into granda-mento
